test/kubernetes/e2e/features/services/tcproute/suite.go

```diff
@@ -31,95 +37,271 @@ func NewTestingSuite(ctx context.Context, testInst *e2e.TestInstallation) suite.
+			svcManifest:      multiSvcBackendManifest,
+			tcpRouteManifest: multiSvcTcpRouteManifest,
+			proxyService:     multiProxyService,
+			proxyDeployment:  multiProxyDeployment,
+			expectedResponses: []*matchers.HttpResponse{
+				expectedMultiSvc1Resp,
+				expectedMultiSvc2Resp,
+			},
+			ports: []int{8088, 8089},
+			listenerNames: []v1.SectionName{
+				v1.SectionName(multiSvcListenerName8088),
+				v1.SectionName(multiSvcListenerName8089),
+			},
+			expectedRouteCounts: []int32{1, 1},
+			tcpRouteNames:       []string{multiSvcTCPRouteName1, multiSvcTCPRouteName2},
 		},
-		expectedTcpFooSvcResp)
-	s.testInstallation.Assertions.AssertEventualCurlResponse(
-		s.ctx,
-		defaults.CurlPodExecOpt,
-		[]curl.Option{
-			curl.WithHost(kubeutils.ServiceFQDN(proxyService.ObjectMeta)),
-			curl.WithPort(8089),
+		{
+			name:             "CrossNamespaceTCPRouteWithReferenceGrant",
+			nsManifest:       crossNsClientNsManifest,
+			gtwName:          crossNsGatewayName,
+			gtwNs:            crossNsClientName,
+			gtwManifest:      crossNsGatewayManifest,
+			svcManifest:      crossNsBackendSvcManifest,
+			tcpRouteManifest: crossNsTCPRouteManifest,
+			proxyService:     crossNsProxyService,
+			proxyDeployment:  crossNsProxyDeployment,
+			expectedResponses: []*matchers.HttpResponse{
+				expectedCrossNsResp,
+			},
+			ports: []int{8080},
+			listenerNames: []v1.SectionName{
+				v1.SectionName(crossNsListenerName),
+			},
+			expectedRouteCounts: []int32{1},
+			tcpRouteNames:       []string{crossNsTCPRouteName},
 		},
-		expectedTcpBarSvcResp)
+		{
+			name:              "CrossNamespaceTCPRouteWithoutReferenceGrant",
+			nsManifest:        crossNsNoRefGrantClientNsManifest,
+			gtwName:           crossNsNoRefGrantGatewayName,
+			gtwNs:             crossNsNoRefGrantClientName,
+			gtwManifest:       crossNsNoRefGrantGatewayManifest,
+			svcManifest:       crossNsNoRefGrantBackendSvcManifest,
+			tcpRouteManifest:  crossNsNoRefGrantTCPRouteManifest,
+			proxyService:      crossNsNoRefGrantProxyService,
+			proxyDeployment:   crossNsNoRefGrantProxyDeployment,
+			expectedErrorCode: 7,
+			ports:             []int{8080},
+			listenerNames: []v1.SectionName{
+				v1.SectionName(crossNsNoRefGrantListenerName),
+			},
+			expectedRouteCounts: []int32{1},
+			tcpRouteNames:       []string{crossNsNoRefGrantTCPRouteName},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc // capture range variable
+		s.Run(tc.name, func() {
+			// Cleanup function
+			s.T().Cleanup(func() {
+				err := s.deleteManifests(tc.nsManifest)
+				s.Require().NoError(err, fmt.Sprintf("Failed to delete manifest %s", tc.nsManifest))
+
+				// Delete additional namespaces if any
+				if tc.name == "CrossNamespaceTCPRouteWithReferenceGrant" {
+					err = s.deleteManifests(crossNsBackendNsManifest)
+					s.Require().NoError(err, fmt.Sprintf("Failed to delete manifest %s", crossNsBackendNsManifest))
+				}
+
+				if tc.name == "CrossNamespaceTCPRouteWithoutReferenceGrant" {
```

Comment:
nit: We use this string alot, perhaps turn it into a constant?

Reply:
Resolved in commit https://github.com/solo-io/gloo/commit/281ace000966aa7f15d960ab73d60efc4fa1c189 by adding string const for this test case name.